executor/set_test.go

```diff
@@ -242,6 +247,18 @@ func (s *testSerialSuite1) TestSetVar(c *C) {
 	tk.MustQuery("select @@global.tx_read_only").Check(testkit.Rows("1"))
 	tk.MustQuery("select @@global.transaction_read_only").Check(testkit.Rows("1"))
 
+	_, err = tk.Exec("SET tidb_enable_noop_functions = 0") // fails because transaction_read_only/tx_read_only = 1
+	c.Assert(err, NotNil)
+
+	tk.MustExec("SET transaction_read_only = 0")
+	tk.MustExec("SET tidb_enable_noop_functions = 0") // now works.
+
+	// setting session doesn't change global, which also fails because transaction_read_only/tx_read_only = 1
```

Comment:
For (1) - I can't reproduce this:
```sql
mysql> set global transaction_read_only = 1;
ERROR 1235 (42000): function READ ONLY has only noop implementation in tidb now, use tidb_enable_noop_functions to enable these functions
```

I think the problem is caused because a previous test already changed the global to ON. I will improve the tests so that they clean up after themselves correctly.

For (2), this is expected behavior. See the comment by @jackysp w.r.t. the logic of validating `TiDBEnableNoopFunctions`. If noop functions is allowed to be disabled, then there is an inconsistent state ("read only" is permitted, but it is not supposed to be.)

Reply:
Nevermind. I found the bug: 
```sql
mysql> set global transaction_read_only = 1;
ERROR 1235 (42000): function READ ONLY has only noop implementation in tidb now, use tidb_enable_noop_functions to enable these functions
mysql> set session tx_read_only = 1;
ERROR 1235 (42000): function READ ONLY has only noop implementation in tidb now, use tidb_enable_noop_functions to enable these functions
mysql> set tidb_enable_noop_functions=1;
Query OK, 0 rows affected (0.00 sec)

mysql> set session tx_read_only = 1;
Query OK, 0 rows affected (0.00 sec)

mysql> set global transaction_read_only = 1;
Query OK, 0 rows affected (0.00 sec)
```

Reply:
I thought about it a little bit, and changed the behavior:
- When setting SESSION scope, it considers the SESSION value of `tidb_enable_noop_functions`.
- When setting GLOBAL scope, it considers the GLOBAL value of `tidb_enable_noop_functions`.

This has been addressed in https://github.com/pingcap/tidb/pull/23818/commits/62738a0834162e2b8fd59d3bed5e6107d358a807 - the testcases have all been grouped together to handle the behavior change.